File: server/src/main/java/xyz/chener/ext/napt/server/core/ClientManager.java
package xyz.chener.ext.napt.server.core;

/**
 * @Author: chenzp
 * @Date: 2023/05/22/17:23
 * @Email: chen@chener.xyz
 */

import com.baomidou.mybatisplus.extension.conditions.query.LambdaQueryChainWrapper;
import com.baomidou.mybatisplus.extension.conditions.update.LambdaUpdateChainWrapper;
import io.netty.channel.ChannelHandlerContext;
import xyz.chener.ext.napt.server.core.requestNt.RequestNtTcp;
import xyz.chener.ext.napt.server.entity.ClientItem;
import xyz.chener.ext.napt.server.entity.DataFrameCode;
import xyz.chener.ext.napt.server.entity.DataFrameEntity;
import xyz.chener.ext.napt.server.mapper.ClientItemMapper;

import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;

/**
 * 刷新 web 在线修改后的client连接
 */
public class ClientManager {

    private final Object lock = new Object();

    public void flush(String clientUid){
        if (!ConnectCache.clientChannel.containsKey(clientUid)) {
            return;
        }
        List<RequestNtTcp> requestNtTcps = ConnectCache.portStarts.get(clientUid);

        List<ClientItem> clients = new LambdaQueryChainWrapper<>(StrongStarter.getMapper(ClientItemMapper.class))
                .eq(ClientItem::getClientUid, clientUid).list();

        // step1 requestNts有 clients 没有的  requestNts删除

        Iterator<RequestNtTcp> it = requestNtTcps.stream().filter(requestNtTcp -> clients.stream().noneMatch(c -> c.getClientUid().equals(requestNtTcp.getClientUid())
                && c.getServerPort().equals(requestNtTcp.getPort())
                && c.getClientAddr().equals(requestNtTcp.getClientAddr()))).iterator();

        while (it.hasNext()){
            RequestNtTcp requestNtTcp = it.next();
            requestNtTcp.stop();
            it.remove();
        }

        // step2 clients有 requestNtTcps 没有的  requestNts添加
        ArrayList<ClientItem> clientSub = new ArrayList<>(clients);
        clientSub.removeIf(ct -> requestNtTcps.stream().anyMatch(requestNtTcp -> requestNtTcp.getClientAddr().equals(ct.getClientAddr()) && requestNtTcp.getPort().equals(ct.getServerPort()) && requestNtTcp.getClientUid().equals(ct.getClientUid())));
        clientSub.forEach(e->{
            requestNtTcps.add(new RequestNtTcp(e.getClientUid(),e.getServerPort(),e.getClientAddr(), e.getSpeedLimit().intValue()));
        });

        // step3 requestNts为空的话全部删除
        if (requestNtTcps.isEmpty()){
            ConnectCache.portStarts.remove(clientUid);
            String channelId = ConnectCache.clientChannel.remove(clientUid);
            ChannelHandlerContext ctx = ConnectCache.channelMap.remove(channelId);
            if (ctx != null){
                ctx.channel().writeAndFlush(DataFrameEntity.DataFrame.newBuilder()
                        .setCode(DataFrameCode.ACCESS_FAIL)
                        .setMessage("客户端ID未绑定任何节点")
                        .build());
                ctx.channel().close();
            }
        }
    }

    public boolean addClient (String clientUid,String clientAddr,int port,Long maxFlowLimit,Long speedLimit,String remark){
        synchronized (lock){
            ClientItemMapper clientItemMapper = StrongStarter.getMapper(ClientItemMapper.class);
            Long count = new LambdaQueryChainWrapper<>(clientItemMapper)
                    .eq(ClientItem::getClientUid, clientUid)
                    .eq(ClientItem::getClientAddr, clientAddr)
                    .eq(ClientItem::getServerPort, port)
                    .count();
            if (count > 0)
                return false;

            ClientItem ci = new ClientItem();
            ci.setClientUid(clientUid);
            ci.setClientAddr(clientAddr);
            ci.setServerPort(port);
            ci.setCreateTime(new Date());
            ci.setMaxFlowLimit(maxFlowLimit);
            ci.setSpeedLimit(speedLimit);
            ci.setRemark(remark);
            clientItemMapper.insert(ci);
            flush(clientUid);
            return true;
        }
    }



    public void removeClient (String clientUid,String clientAddr,int port){
        synchronized (lock){
            ClientItemMapper clientItemMapper = StrongStarter.getMapper(ClientItemMapper.class);
            new LambdaUpdateChainWrapper<>(clientItemMapper)
                    .eq(ClientItem::getClientUid, clientUid)
                    .eq(ClientItem::getClientAddr, clientAddr)
                    .eq(ClientItem::getServerPort, port).remove();
            flush(clientUid);
        }
    }



}

File: server/src/main/kotlin/xyz/chener/napt/server/core/ClientManager.kt
package xyz.chener.napt.server.core

import io.netty.channel.ChannelHandlerContext
import org.springframework.stereotype.Component
import xyz.chener.napt.common.entity.DataFrameCode
import xyz.chener.napt.common.entity.DataFrameEntity
import xyz.chener.napt.common.entity.ProxyType
import xyz.chener.napt.server.core.proxy.AbstractPortProxy
import xyz.chener.napt.server.entity.ClientItem
import xyz.chener.napt.server.http.entity.SyncLockPayload
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList


@Component
open class ClientManager {


    // clientUid --> clientChannelId
    val clientUidToClientChannelId: ConcurrentHashMap<String, String> = ConcurrentHashMap()


    // channelId -> ChannelHandlerContext
    val channelIdToChannel: ConcurrentHashMap<String, ChannelHandlerContext> = ConcurrentHashMap()


    // clientUid -> 端口代理   存储每个客户端的转发服务的服务端port列表
    val portProxys: ConcurrentHashMap<String, CopyOnWriteArrayList<AbstractPortProxy>> = ConcurrentHashMap<String, CopyOnWriteArrayList<AbstractPortProxy>>()

    // 同步获取后端信息  使用的等待器  供HTTP部分使用
    val getClientInfoLockCache = ConcurrentHashMap<String, SyncLockPayload>()

    /**
     * 后端客户端连接断开时执行的清理方法
     */
    fun onClientClose(clientChannelId :String){
        val clientUid = findClientUidByClientChannelId(clientChannelId)
        clientUid?.let {
            clientUidToClientChannelId.remove(it)

            // 把所有监听的代理关闭..
            portProxys[clientUid]?.forEach { it0 ->
                it0.runCatching {
                    this.stop()
                }
            }

            portProxys[clientUid]?.clear()
            portProxys.remove(clientUid)
        }
        channelIdToChannel.remove(clientChannelId)
    }


    /**
     * 踢出客户端
     */
    fun kickClient(clientUid: String){
        val channelId = clientUidToClientChannelId[clientUid]

        channelId?.let {
            val channelHandlerContext = channelIdToChannel[channelId]
            channelHandlerContext?.channel()?.writeAndFlush(
                DataFrameEntity.DataFrame.newBuilder()
                .setCode(DataFrameCode.CLOSE.code)
                .setMessage("${DataFrameCode.CLOSE.message}:其它客户端登录,强制断开连接")
                .build())
            channelHandlerContext?.channel()?.close()
            onClientClose(channelId)
        }
    }


    /**
     * 授权客户端
     */
    fun authClient(context : ChannelHandlerContext,clientUid: String,clientItems:List<ClientItem>){

        val message = StringBuilder().also { str ->
             clientItems.forEach {
                 str.append("IP:").append(it.serverPort).append(" --> ").append(it.clientAddr).append("\n")
             }
        }

        context.channel().writeAndFlush(
            DataFrameEntity.DataFrame.newBuilder()
                .setCode(DataFrameCode.ACCESS_SUCCESS.code)
                .setMessage(message.toString())
                .build()
        )

        clientUidToClientChannelId[clientUid] = context.channel().id().asLongText()
        channelIdToChannel[context.channel().id().asLongText()] = context

        val proxys = CopyOnWriteArrayList<AbstractPortProxy>().also { proxy ->
            // 启动代理端口
            clientItems.forEach {
                val proxyType = when(it.requestNtType){
                    1 -> ProxyType.TCP
                    2 -> ProxyType.UDP
                    else -> throw Exception("未知的代理类型:${it.requestNtType}")
                }
                val portProxy = AbstractPortProxy.createProxy(proxyType,clientUid,it.serverPort!!,it.clientAddr!!,it.speedLimit!!.toInt())
                proxy.add(portProxy)
            }
        }

        portProxys[clientUid] = proxys
    }

    private fun findClientUidByClientChannelId(clientChannelId: String): String? {
        return clientUidToClientChannelId.entries.find { it.value == clientChannelId }?.key
    }
}